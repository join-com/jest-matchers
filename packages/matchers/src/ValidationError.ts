export interface IFieldError {
  fieldName: string
  message: string
  constraint: string
}

export class ValidationError extends Error {
  public readonly type = 'ApplicationError'
  public readonly code = 'validation'

  constructor(
    message: string,
    public readonly fields: IFieldError[],
  ) {
    super(message)
  }
}
