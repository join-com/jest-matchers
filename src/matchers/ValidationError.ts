export class ValidationError extends Error {
  public readonly type = 'ApplicationError';
  public readonly code = 'validation';

  constructor(
    message: string,
    public readonly fields: {
      fieldName: string;
      message: string;
      constraint: string;
    }[],
  ) {
    super(message);
  }
}
