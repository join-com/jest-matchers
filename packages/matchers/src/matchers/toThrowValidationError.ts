import equal from 'fast-deep-equal'
import { printDiffOrStringify } from 'jest-matcher-utils'
import { ValidationError } from '../ValidationError'

function getThrown(received: Error | (() => void)): Error | undefined {
  if (received instanceof Error) {
    return received
  }

  try {
    received()
  } catch (error) {
    return error as Error
  }

  return
}

export function toThrowValidationError<T extends ValidationError>(
  received: Error | (() => void),
  expected: T,
): jest.CustomMatcherResult {
  const thrown = getThrown(received)
  if (!thrown) {
    return {
      pass: false,
      message: () => 'Expected to receive an error or a function that throws an error',
    }
  }

  const isExpectedClass = thrown instanceof expected.constructor
  if (!isExpectedClass) {
    return {
      pass: false,
      message: () =>
        `Expected error to be an instance of ${expected.constructor.name} but got ${thrown.constructor.name}`,
    }
  }

  const receivedFields = (thrown as ValidationError).fields
  const areFieldsMatching = equal(receivedFields, expected.fields)
  if (!areFieldsMatching) {
    return {
      pass: false,
      message: () => printDiffOrStringify(receivedFields, expected.fields, 'Expected fields', 'Received fields', true),
    }
  }

  return { pass: true, message: () => '' }
}
