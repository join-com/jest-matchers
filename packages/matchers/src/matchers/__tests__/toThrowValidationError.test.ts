import { IFieldError, ValidationError } from '../../ValidationError'
import { toThrowValidationError } from '../toThrowValidationError'

expect.extend({ toThrowValidationError })

describe('toThrowValidationError', () => {
  const expectedValidationError = (partialFieldError?: Partial<IFieldError>) =>
    new ValidationError('Expected validation error', [
      {
        fieldName: 'someField',
        constraint: 'mustBeCorrect',
        message: 'Some field must be correct but its not',
        ...partialFieldError,
      },
    ])

  it('passes for function throwing expected ValidationError', async () => {
    const fn = () => {
      throw expectedValidationError()
    }

    await expect(fn).toThrowValidationError(expectedValidationError())
  })

  it('passes for directly passing the expected ValidationError', async () => {
    await expect(expectedValidationError()).toThrowValidationError(expectedValidationError())
  })

  it('fails when error is not thrown', async () => {
    const fn = () => {
      /** Do nothing */
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for error of other type', async () => {
    const fn = () => {
      throw new Error('other type')
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for ValidationError with different fieldName', async () => {
    const fn = () => {
      const error = expectedValidationError({ fieldName: 'different' })
      throw error
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for ValidationError with different constraint', async () => {
    const fn = () => {
      const error = expectedValidationError({ constraint: 'different' })
      throw error
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }
    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for ValidationError with different message', async () => {
    const fn = () => {
      const error = expectedValidationError({ message: 'different' })
      throw error
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for ValidationError with additional fields', async () => {
    const fn = () => {
      const error = expectedValidationError()
      error.fields[1] = {
        fieldName: 'someOtherField',
        constraint: 'mustBeCorrectToo',
        message: 'Some other field must be correct but its not',
      }
      throw error
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for ValidationError with no fields', async () => {
    const fn = () => {
      const error = expectedValidationError()
      error.fields.pop()
      throw error
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })

  it('fails for ValidationError with different fields', async () => {
    const fn = () => {
      const error = expectedValidationError()
      error.fields[0] = {
        fieldName: 'someOtherField',
        constraint: 'mustBeCorrectToo',
        message: 'Some other field must be correct but its not',
      }
      throw error
    }

    const validate = async () => {
      await expect(fn).toThrowValidationError(expectedValidationError())
    }

    await expect(validate).rejects.toThrowErrorMatchingSnapshot()
  })
})
