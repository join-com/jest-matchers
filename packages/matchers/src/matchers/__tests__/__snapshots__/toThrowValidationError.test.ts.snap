// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toThrowValidationError fails for ValidationError with additional fields 1`] = `
"- Expected fields  - 5
+ Received fields  + 0

  Array [
    Object {
      "constraint": "mustBeCorrect",
      "fieldName": "someField",
      "message": "Some field must be correct but its not",
    },
-   Object {
-     "constraint": "mustBeCorrectToo",
-     "fieldName": "someOtherField",
-     "message": "Some other field must be correct but its not",
-   },
  ]"
`;

exports[`toThrowValidationError fails for ValidationError with different constraint 1`] = `
"- Expected fields  - 1
+ Received fields  + 1

  Array [
    Object {
-     "constraint": "different",
+     "constraint": "mustBeCorrect",
      "fieldName": "someField",
      "message": "Some field must be correct but its not",
    },
  ]"
`;

exports[`toThrowValidationError fails for ValidationError with different fieldName 1`] = `
"- Expected fields  - 1
+ Received fields  + 1

  Array [
    Object {
      "constraint": "mustBeCorrect",
-     "fieldName": "different",
+     "fieldName": "someField",
      "message": "Some field must be correct but its not",
    },
  ]"
`;

exports[`toThrowValidationError fails for ValidationError with different fields 1`] = `
"- Expected fields  - 3
+ Received fields  + 3

  Array [
    Object {
-     "constraint": "mustBeCorrectToo",
-     "fieldName": "someOtherField",
-     "message": "Some other field must be correct but its not",
+     "constraint": "mustBeCorrect",
+     "fieldName": "someField",
+     "message": "Some field must be correct but its not",
    },
  ]"
`;

exports[`toThrowValidationError fails for ValidationError with different message 1`] = `
"- Expected fields  - 1
+ Received fields  + 1

  Array [
    Object {
      "constraint": "mustBeCorrect",
      "fieldName": "someField",
-     "message": "different",
+     "message": "Some field must be correct but its not",
    },
  ]"
`;

exports[`toThrowValidationError fails for ValidationError with no fields 1`] = `
"- Expected fields  - 1
+ Received fields  + 7

- Array []
+ Array [
+   Object {
+     "constraint": "mustBeCorrect",
+     "fieldName": "someField",
+     "message": "Some field must be correct but its not",
+   },
+ ]"
`;

exports[`toThrowValidationError fails for error of other type 1`] = `"Expected error to be an instance of ValidationError but got Error"`;

exports[`toThrowValidationError fails when error is not thrown 1`] = `"Expected to receive an error or a function that throws an error"`;
